stan/math/rev/arr/functor/map_rect_lpdf_mpi.hpp

```diff
@@ -0,0 +1,336 @@
+#pragma once
+
+#include <vector>
+#include <set>
+
+#include <boost/mpi/operations.hpp>
+
+#include <stan/math/prim/mat/fun/to_array_1d.hpp>
+#include <stan/math/prim/arr/functor/mpi_command.hpp>
+#include <stan/math/prim/arr/functor/map_rect_mpi.hpp>
+//#include <stan/math/rev/arr/functor/map_rect_mpi.hpp>
+
+namespace stan {
+  namespace math {
+
+    namespace internal {
+      
+      template <typename F>
+      class distributed_map_rect_lpdf {
+        boost::mpi::communicator world_;
+        const std::size_t R_ = world_.rank();
+        const std::size_t W_ = world_.size();
+
+        // local version of things in flattened out form
+        std::vector<double> local_eta_;
+        std::vector<double> local_theta_;
+        std::size_t uid_;
+
+        // note: it would be nice to have these const, but I think C++
+        // doesn't let me given how it has to be initialized
+        std::size_t N_; // # of jobs for world
+        std::size_t E_; // # of shared parameters
+        std::size_t T_; // # of parameters
+        
+        // # of outputs per job
+        double *final_result_ = 0;
+
+        // the parameters which can be updated; only stored at the
+        // root
+        const std::vector<var>* eta_;
+        typedef const std::vector<std::vector<var> > param_t;
+        param_t* theta_;
+
+        std::vector<int> chunks_;
+
+        // # of jobs for this chunk
+        std::size_t C_;
+
+      public:
+        // called on root, sents problem sizes, data and parameters
+        distributed_map_rect_lpdf(const std::vector<var>& eta,
```

Comment:
Can you make use of and provide a more general API similar to `boost::mpi::reduce`: https://theboostcpplibraries.com/boost.mpi-collective-data-exchange#ex.mpi_13 ?

Reply:
I have used reduce in my previous version (just go back in the history). Benchmarks did show me that it was slower than the mpi_rect which does not even do a sum. On the net I found documents which make me question if mpi::reduce is a good thing to use. Also, what I found out is that we should absolutely make sure that we use the C MPI calls after all. That is, using abstractions from boost::mpi is certainly something we do not want as its super slow. In short, I abandoned that approach with mpi::reduce and performance seems to get a little better or stay the same, don't know. Wer are anyway going to do the mpi_rect_lpdf as a second pull as I understood, so let's worry about this later?

Reply:
Sure, we can worry about the `mpi_rect_lpdf` later.